src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -7752,6 +7752,295 @@ void M()
+  IL_000a:  ldloca.s   V_1
+  IL_000c:  ldloca.s   V_2
+  IL_000e:  call       ""void Extension.Deconstruct(in S, out int, out int)""
+  IL_0013:  nop
+  IL_0014:  ldarg.1
+  IL_0015:  ldloc.1
+  IL_0016:  stind.i4
+  IL_0017:  ldarg.2
+  IL_0018:  ldloc.2
+  IL_0019:  stind.i4
+  IL_001a:  ret
+}");
+        }
+
+        [Fact, WorkItem(24160, "https://github.com/dotnet/roslyn/issues/24160")]
+        public void TestInThisDeconstruct_Nested()
+        {
+            var source = @"
+public class C
+{
+    static S field = default;
+    static void Main()
+    {
+        var (x, (y, z)) = (1, field);
+        System.Console.Write($""{y} {z}"");
+    }
+}
+public struct S { }
+public static class Extension
+{
+    public static void Deconstruct(in this S d, out int x1, out int x2)
+    {
+        x1 = 1;
+        x2 = 2;
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "1 2");
+        }
+
+        [Fact, WorkItem(24160, "https://github.com/dotnet/roslyn/issues/24160")]
+        public void TestInThisDeconstruct_Foreach()
+        {
+            var source = @"
+public class C
+{
+    static S field = default;
+    static void Main()
+    {
+        foreach (var (y, z) in new[] { field })
+        {
+            System.Console.Write($""{y} {z}"");
+        }
+    }
+}
+public struct S { }
+public static class Extension
+{
+    public static void Deconstruct(in this S d, out int x1, out int x2)
+    {
+        x1 = 1;
+        x2 = 2;
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "1 2");
+        }
+
+        [Fact, WorkItem(24160, "https://github.com/dotnet/roslyn/issues/24160")]
+        public void TestRefThisDeconstruct()
+        {
+            var source = @"
+public class C
+{
+    S field = default;
+    static void Main()
+    {
+        new C().M();
+    }
+    void M()
+    {
+        var (x, y) = field;
+        System.Console.Write($""{field.touched} {x + y}"");
+    }
+}
+public struct S
+{
+    public bool touched;
+}
+public static class Extension
+{
+    public static void Deconstruct(ref this S d, out int x1, out int x2)
+    {
+        // d is a reference to a temp used for deconstruction
+        d.touched = true;
+        System.Console.Write(""Deconstruct "");
+        x1 = 1;
+        x2 = 2;
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "Deconstruct False 3");
+        }
+
+        [Fact, WorkItem(24160, "https://github.com/dotnet/roslyn/issues/24160")]
+        public void TestRefThisDeconstruct_Nested()
```

Comment:
Do we have a test where there is an attempt to deconstruct a readonly variable? 
Deconstruct with "this in" should be ok (no copies).
Deconstruct with "this ref" should be an error.